### PR TITLE
list exceptions from the doc comments when showing quick info

### LINKED ISSUE #4868: QuickInfo does not list exceptions from documentation comments.
When invoking QuickInfo on the `StreamReader` constructor:

``` C#
var reader = new StreamReader("");
```

VS 2013 gives the following QuickInfo:

    StreamReader.StreamReader(string path) (+ 10 overloads)
    Initializes a new instance of the StreamReader class for the specified file name.
    
    Exceptions:
      System.ArgumentException
      System.ArgumentNullException
      System.IO.DirectoryNotFoundException
      System.IO.FileNotFoundException
      System.IO.IOException

But VS 2015 only gives:

    StreamReader.StreamReader(string path) (+ 10 overloads)
    Initializes a new instance of the StreamReader class for the specified file name.

(Issue was reported by a customer and tracked as internal bug 1213849.)

### PR DESCRIPTION
**Pull request by brettfo**
This fixes a TODO comment in the Quick Info code and also addresses a customer-filed bug; VS2013 displays exception info, but VS2015 does not.

The bulk of this PR is simple plumbing with the real work occurring in `AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs `.

Tagging @Pilchie @dpoeschl @jasonmalinowski @rchande @balajikris @CyrusNajmabadi @jmarolf @davkean for review.

_Preview of how it looks now:_

Fixes #4868.

### CODE REVIEW

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs [hunk 1]
```diff
@@ -151,6 +151,38 @@ protected SemanticModel GetSemanticModel(SyntaxTree tree)
...
+                // prefix before it is trimmed off and parsed.  All strings that don't parse are simply
+                // ignored
+                var exceptions =
+                    symbol.GetDocumentationComment().ExceptionTypes
+                    .Distinct()
+                    .Where(e => e.StartsWith("T:"))
+                    .Select(e => e.Substring(2))
+                    .OrderBy(e => e)
+                    .Select(_semanticModel.Compilation.GetTypeByMetadataName)
+                    .WhereNotNull();
```
Comment (brettfo) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: I'm not sure if there is a better way to get symbols for the exception types than what I'm doing here.  `.ExceptionTypes` only returns strings which then have to be parsed.

I'm certainly open to suggestions for changing this if need be.

**Comment by davkean** [comment]
Can you attach a screenshot?

**Comment by brettfo** [comment]
@davkean Modified the original post with a sample.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs (hunk 1 above)
Comment (rchande) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: I vaguely remember that we have code to do this, but I can't remember the name now...

File: src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb [hunk 2]
```diff
@@ -2032,5 +2032,32 @@ End Interface
...
+        ''' <exception cref=""System.Int32""></exception>
+        ''' <exception cref=""Double""></exception>
+        ''' <exception cref=""Not_A_Class_And_Therefore_Not_Listed""></exception>
+        Sub M()
+            M$$()
+        End Sub
+    End Class
+End Namespace
+",
+                Exceptions($"{vbCrLf}{WorkspacesResources.Exceptions}{vbCrLf}  MyException1{vbCrLf}  MyException2{vbCrLf}  Double{vbCrLf}  Integer"))
```
Comment (rchande) [inline src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb]: Perhaps tests that include exception info as well as the "awaitable" text to verify those come out in the right order?

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs (hunk 1 above)
Comment (rchande) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: DocumentationCommentId.GetFirstSymbolForDeclarationId

**Comment by rchande** [comment]
:+1:

**Comment by davkean** [comment]
This looks good, but please don't consider this a senior signoff - still figuring out the code base.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs (hunk 1 above)
Comment (davkean) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: One more thing, you're calculating this twice for the first exception - once for the Any call, and then once again for the foreach.

Comment (brettfo) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: Are the results not cached and re-used?  If they're not then I'll force realization via `.ToList()` or similar.

Comment (davkean) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: None of the Linq-provided methods above cache, and Roslyn's WhereNotNul doesn't.

File: src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb (hunk 2 above)
Comment (brettfo) [inline src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb]: Unfortunately this isn't easily done.  The final display is all done via a `StackPanel` in `QuickInfoDisplayPanel.cs`, so the best we could do is by using the `.ToString()` overload which happens to generate the text in the same order as the `StackPanel`, but there's no guarantee that these two things (the order or the `StackPanel` and the order of evaluation in `.ToString()`) will stay in sync.

To make it even more complicated, `.ToString()` eventually evaluates the `System.Windows.Documents.Run::Text` property which inserts left-to-right Unicode markers, and while we could explicitly add those in the test, as soon as we run our tests on an RTL language then we're broken again.

For what it's worth, I did some manual testing with exceptions and await (which generates the `Usage:` section) and the display is consistent with VS2013.

**tl;dr;**
Not worth it.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs (hunk 1 above)
Comment (brettfo) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: Thanks for the clarification; I'll update to prevent duplicate realization.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs [hunk 3]
```diff
@@ -151,6 +151,33 @@ protected SemanticModel GetSemanticModel(SyntaxTree tree)
...
+                AddExceptions(symbols[0]);
+            }
+
+            private void AddExceptions(ISymbol symbol)
+            {
+                // clean up the list of possible exceptions by de-duplicating and ordering them
+                var exceptions =
+                    symbol.GetDocumentationComment().ExceptionTypes
+                    .Distinct()
+                    .OrderBy(e => e)
```
Comment (sharwell) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: :exclamation: If these are strings, you need to specify `StringComparer.____IgnoreCase` (the `____` would be up to other reviewers).

File: src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb (hunk 2 above)
Comment (rchande) [inline src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb]: As it turns out, integration tests know how to do that. Checkout `QuickInfoToStringConverter`, if you're interested.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs (hunk 3 above)
Comment (davkean) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: That's a good point. Distinct itself is going to do a ordinal filter - which is correct. 

Given this is for display purposes, you should pass in StringComparer.CurrentCultureIgnoreCase OrderBy to order them in the current culture's sort order.

Comment (brettfo) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: Then this brings up the issue of testability again.  If we use the current culture then we could presumably get a different sort order when running unit tests on other languages and that's a maintenance nightmare (coming from the person who had to make all of these tests run on various cultures a few months ago.)

A question for @Pilchie and/or @DustinCampbell: what comparer should I specify for the sort, or should we fall back to not sorting anything and displaying the exceptions in the order in which they were declared in the original doc comment?

File: src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb (hunk 2 above)
Comment (brettfo) [inline src/EditorFeatures/VisualBasicTest/QuickInfo/SemanticQuickInfoSourceTests.vb]: Thanks for the pointer.  I've submitted a corresponding internal PR to test this.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs [hunk 4]
```diff
@@ -151,6 +151,33 @@ protected SemanticModel GetSemanticModel(SyntaxTree tree)
...
+                    .Select(e => DocumentationCommentId.GetFirstSymbolForDeclarationId(e, _semanticModel.Compilation))
+                    .WhereNotNull()
+                    .ToList();
+                if (exceptions.Any())
+                {
+                    var parts = new List<SymbolDisplayPart>();
+                    parts.Add(new SymbolDisplayPart(SymbolDisplayPartKind.Text, null, $"\r\n{WorkspacesResources.Exceptions}"));
```
Comment (balajikris) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: nit: consider named arguments when using default values

File: src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs [hunk 5]
```diff
@@ -4135,5 +4135,32 @@ public class Methods2
...
+        /// <exception cref=""Not_A_Class_And_Therefore_Not_Listed""></exception>
+        void M()
+        {
+            M$$();
+        }
+    }
+}
+",
+                Exceptions($"\r\n{WorkspacesResources.Exceptions}\r\n  MyException1\r\n  MyException2\r\n  double\r\n  int"));
+        }
```
Comment (balajikris) [inline src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs]: I pinged @brettfo that the sort order looks odd here.. He explains that we sort using fully qualified names (MyNs.MyEx, System.Double) and then symbol display uses a simplified name, thus making it look unsorted.

I think if we're taking the time to sort the list, we should ensure it looks sorted. If its going to look unsorted, we should atleast save the time used for sorting and not sort at all  :smile:

Comment (balajikris) [inline src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs]: What does the prior versions of VS do here?

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs (hunk 3 above)
Comment (balajikris) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: I like the neat sorted order. Also helps me easily spot the exception i'm looking for.
What does VS2013 do? (sorted fully qualified?)

File: src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs (hunk 5 above)
Comment (brettfo) [inline src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs]: VS 2013 simply displays the types as listed in the doc comment, e.g.,

    MyException1
    MyNs.MyException2,
    System.Int32,
    double
    Not_A_Class_And_Therefore_Not_Listed

Having thought about the difficulties in dealing with sorting the fully-qualified names vs. sorting the simplified names (which could change with each new `using`), I'm leaning towards no sorting at all and simply using the declaration order, but still letting the symbol display service simplify names where possible, e.g.,

    MyException1
    MyException2
    int
    double

But there's still the issue of what to do with exceptions that didn't parse as a type.  The doc comment author thought it important to include an exception type here, but why?

Comment (davkean) [inline src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs]: By “parse as a type”, what do you mean? The type didn’t exist in the compilation? If so, I can see that occurring pretty easy – the callee throws an XamlException but don’t have a reference to System.Xaml.dll. In this case, I would expect it to show up.

File: src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/DeferredContent/QuickInfoDisplayDeferredContent.cs [hunk 6]
```diff
@@ -22,7 +23,8 @@ internal class QuickInfoDisplayDeferredContent : IDeferredQuickInfoContent
             IDeferredQuickInfoContent documentation,
             IDeferredQuickInfoContent typeParameterMap,
             IDeferredQuickInfoContent anonymousTypes,
-            IDeferredQuickInfoContent usageText)
+            IDeferredQuickInfoContent usageText,
+            IDeferredQuickInfoContent exceptionText)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/DeferredContent/QuickInfoDisplayDeferredContent.cs]: This will likely break TypeScript.  @vladima @paulvanbrenk , if this goes through, you'll have to provide an additional empty value for this as part of your quick info result.

**Comment by CyrusNajmabadi** [comment]
All this futzing about sorting seems seriously overwrought.  I'd just not do anything and keep it as it was written originally.

**Comment by davkean** [comment]
@CyrusNajmabadi Yep, coming to that conclusion as well, leave them in the order that they were written by the callee.

**Comment by DustinCampbell** [comment]
@CyrusNajmabadi, @davkean - I'm of the same opinion.

**Comment by brettfo** [comment]
I've removed the sort and now if we can't find the symbol for the exception type we just display the raw text after trying to strip off the leading "T:" prefix.  I did this last part because thinking about @davkean's example of a valid but unreferenced exception, the user would rather just see `System.SomeValidButUnreferencedException` instead of `T:System.SomeValidButUnreferencedException`.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs [hunk 7]
```diff
@@ -151,6 +151,41 @@ protected SemanticModel GetSemanticModel(SyntaxTree tree)
...
+
+                        // try to get the actual exception symbol
+                        var exceptionSymbol = DocumentationCommentId.GetFirstSymbolForDeclarationId(exceptionString, _semanticModel.Compilation);
+                        if (exceptionSymbol != null)
+                        {
+                            parts.AddRange(_displayService.ToMinimalDisplayParts(_semanticModel, _position, exceptionSymbol));
+                        }
+                        else
+                        {
+                            // unable to parse exception symbol, fall back to displaying the raw text after trying to
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: I would expect we already had this logic for crefs, or want to share it. Is there some unification that should happen here?

Comment (brettfo) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: @jasonmalinowski: I found where crefs were getting their "T:" prefix stripped and exposed the method as `internal` so it could be re-used here.  See latest commit.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs [hunk 8]
```diff
@@ -151,6 +152,37 @@ protected SemanticModel GetSemanticModel(SyntaxTree tree)
...
+                        // try to get the actual exception symbol
+                        var exceptionSymbol = DocumentationCommentId.GetFirstSymbolForDeclarationId(exceptionString, _semanticModel.Compilation);
+                        if (exceptionSymbol != null)
+                        {
+                            parts.AddRange(_displayService.ToMinimalDisplayParts(_semanticModel, _position, exceptionSymbol));
+                        }
+                        else
+                        {
+                            // unable to parse exception symbol, fall back to displaying the raw text
+                            var displayText = AbstractDocumentationCommentFormattingService.TrimCrefPrefix(exceptionString);
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs]: Can we refactor further? If you look at the callers of TrimCrefPrefix, we now have three places that go "try to bind symbol, if we do show it, else trim the cref prefix and do something else". I think they can be pulled into one place.

**Comment by brettfo** [comment]
@jasonmalinowski Updated with suggested refactorings.

File: src/Features/Core/Portable/DocumentationCommentFormatting/AbstractDocumentationCommentFormattingService.cs [hunk 9]
```diff
@@ -203,44 +182,36 @@ private static void AppendTextFromNode(FormatterState state, XNode node, Compila
...
         }
 
-        private static void AppendTextFromSeeTag(FormatterState state, XElement element, Compilation compilation)
+        internal static IEnumerable<SymbolDisplayPart> CrefToSymbolDisplayParts(string crefValue, int position, SemanticModel semanticModel, SymbolDisplayFormat format = null)
         {
-            var crefAttribute = element.Attribute("cref");
-
-            if (crefAttribute == null)
+            // first try to parse the symbol
+            if (semanticModel?.Compilation != null)
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/DocumentationCommentFormatting/AbstractDocumentationCommentFormattingService.cs]: How would either semanticModel be null, or worse, the compilation of a semantic model?

**Comment by jasonmalinowski** [comment]
:+1:

File: src/Features/Core/Portable/DocumentationCommentFormatting/AbstractDocumentationCommentFormattingService.cs (hunk 9 above)
Comment (brettfo) [inline src/Features/Core/Portable/DocumentationCommentFormatting/AbstractDocumentationCommentFormattingService.cs]: The `FormatterState.TryAppendSymbol()` method that I removed explicitly checked for a null semantic model so I simply carried that check forward.  As for `Compilation`, I have no idea if that can be null or not, but if it is `DocumentationCommentId.GetFirstSymbolForDeclarationId()` will throw if the compilation was null and I was trying to avoid that, but if a semantic model's compilation can never be null (or if it is then something else catastrophic happened) then I'm ok to turn

```C#
if (semanticModel?.Compilation != null)
```

into

``` C#
if (semanticModel != null) // if the model is valid, then so is the compilation
```

Comment (jasonmalinowski) [inline src/Features/Core/Portable/DocumentationCommentFormatting/AbstractDocumentationCommentFormattingService.cs]: SemanticModels cannot have null compilations.

**Comment by brettfo** [comment]
retest this please

File: src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/DeferredContent/QuickInfoDisplayDeferredContent.cs [hunk 10]
```diff
@@ -13,6 +13,7 @@ internal class QuickInfoDisplayDeferredContent : IDeferredQuickInfoContent
         private readonly IDeferredQuickInfoContent _typeParameterMap;
         private readonly IDeferredQuickInfoContent _anonymousTypes;
         private readonly IDeferredQuickInfoContent _usageText;
+        private readonly IDeferredQuickInfoContent _exceptionText;
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/DeferredContent/QuickInfoDisplayDeferredContent.cs]: Not directly related to this change, but does anyone know why these are typed as `IDeferredQuickInfoContent`, but we always cast them to `ClassifiableQuickInfoContent`?

File: src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/Providers/QuickInfoDisplayPanel.cs [hunk 11]
```diff
@@ -24,13 +25,15 @@ internal class QuickInfoDisplayPanel : StackPanel
             FrameworkElement documentation,
             FrameworkElement typeParameterMap,
             FrameworkElement anonymousTypes,
-            FrameworkElement usageText)
+            FrameworkElement usageText,
+            FrameworkElement exceptionText)
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/Providers/QuickInfoDisplayPanel.cs]: Here to, we require `FrameworkElement`, but then always cast to `TextBlock`?

**Comment by Pilchie** [comment]
:+1:
